cloudinit/ssh_util.py

```diff
@@ -211,32 +213,58 @@ def users_ssh_info(username):
     return (os.path.join(pw_ent.pw_dir, '.ssh'), pw_ent)
 
 
-def extract_authorized_keys(username):
+def render_authorizedkeysfile(value, homedir, username):
+    if value is None:
+        value = "%h/.ssh/authorized_keys"
+    for macro, field in (("%h", homedir), ("%u", username), ("%%", "%")):
+        value = value.replace(macro, field)
+        if not value.startswith("/"):
+            value = os.path.join(homedir, value)
+    return value
+
+
+def extract_authorized_keys(username, sshd_cfg_file=DEF_SSHD_CFG):
     (ssh_dir, pw_ent) = users_ssh_info(username)
-    auth_key_fn = None
+    auth_key_fns = []
     with util.SeLinuxGuard(ssh_dir, recursive=True):
         try:
+            default_authorizedkeys_file = "%h/.ssh/authorized_keys"
+
             # The 'AuthorizedKeysFile' may contain tokens
             # of the form %T which are substituted during connection set-up.
             # The following tokens are defined: %% is replaced by a literal
             # '%', %h is replaced by the home directory of the user being
             # authenticated and %u is replaced by the username of that user.
-            ssh_cfg = parse_ssh_config_map(DEF_SSHD_CFG)
-            auth_key_fn = ssh_cfg.get("authorizedkeysfile", '').strip()
-            if not auth_key_fn:
-                auth_key_fn = "%h/.ssh/authorized_keys"
-            auth_key_fn = auth_key_fn.replace("%h", pw_ent.pw_dir)
-            auth_key_fn = auth_key_fn.replace("%u", username)
-            auth_key_fn = auth_key_fn.replace("%%", '%')
-            if not auth_key_fn.startswith('/'):
-                auth_key_fn = os.path.join(pw_ent.pw_dir, auth_key_fn)
+            ssh_cfg = parse_ssh_config_map(sshd_cfg_file)
+            auth_key_fns = re.split(
+                    r'(?<!\\) ',
+                    ssh_cfg.get("authorizedkeysfile", '').strip())
+
+            if not auth_key_fns:
+                auth_key_fns[0] = default_authorizedkeys_file
+            elif len(auth_key_fns) > 1:
+                util.logexc(
+                        LOG,
+                        "Looks like there's more than one authorizedkeys file"
+                        " configured. Make sure there's no white spaces in"
+                        " their paths. If they do, escape with '\\'.")
```

Comment:
I don't understand the error here, can you elaborate on why we're logging this?

Reply:
The problem here is that sshd_config file doesn't treat white spaces. So for example, something like this:
`AuthorizedKeysFile /tmp/authorized_keys /tmp/user keys`
would split /tmp/users and keys into two different files. We just need to make sure the user escapes the filenames